overlord/snapstate/backend/link_test.go

```diff
@@ -218,3 +221,95 @@ func (s *linkSuite) TestLinkFailsForUnsetRevision(c *C) {
 	err := s.be.LinkSnap(info)
 	c.Assert(err, ErrorMatches, `cannot link snap "foo" with unset revision`)
 }
+
+// testLinkCleanupDirOnFail is dual-purpose: it's the setup for all
+// TestLinkCleanupOnXFail tests below, and the actual test when dir is
+// non-empty.
+func (s *linkSuite) testLinkCleanupDirOnFail(c *C, dir string) *snap.Info {
+	const yaml = `name: hello
+version: 1.0
+environment:
+ KEY: value
+
+apps:
+ foo:
+   command: foo
+ bar:
+   command: bar
+ svc:
+   command: svc
+   daemon: simple
+`
+	info := snaptest.MockSnap(c, yaml, "", &snap.SideInfo{Revision: snap.R(11)})
+
+	guiDir := filepath.Join(info.MountDir(), "meta", "gui")
+	c.Assert(os.MkdirAll(guiDir, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(filepath.Join(guiDir, "bin.desktop"), []byte(`
+[Desktop Entry]
+Name=bin
+Icon=${SNAP}/bin.png
+Exec=bin
+`), 0644), IsNil)
+
+	systemd.SystemctlCmd = func(...string) ([]byte, error) {
+		return nil, nil
+	}
+
+	// sanity checks
+	for _, d := range []string{dirs.SnapBinariesDir, dirs.SnapDesktopFilesDir, dirs.SnapServicesDir} {
+		os.MkdirAll(d, 0755)
+		l, err := filepath.Glob(filepath.Join(d, "*"))
+		c.Assert(err, IsNil, Commentf(d))
+		c.Assert(l, HasLen, 0, Commentf(d))
+	}
+
+	if dir == "" {
```

Comment:
// just invoked for setup   ?

otoh I find this double usage even if commented a bit strange